File: javascript/TarsaLZP/js/core/Encoder.js
/*
 * Copyright (c) 2012, Piotr Tarsa
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * Neither the name of the author nor the names of its contributors may be used
 * to endorse or promote products derived from this software without specific
 * prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
function newEncoder(inputStream, outputStream, options) {
    var rcBuffer = 0;
    var rcRange = 0x7fffffff;
    var xFFRunLength = 0;
    var lastOutputByte = 0;
    var delay = false;
    var carry = false;

    var _super = newCommon(options);
    var self = Object.create(_super);

    function _outputByte(octet) {
        if (octet != 0xff) {
            if (delay) {
                outputStream.write(lastOutputByte + (carry ? 1 : 0));
            }
            while (xFFRunLength > 0) {
                xFFRunLength--;
                outputStream.write(carry ? 0x00 : 0xff);
            }
            lastOutputByte = octet;
            delay = true;
            carry = false;
        } else {
            xFFRunLength++;
        }
    }

    function _normalize() {
        while (rcRange < 0x00800000) {
            _outputByte((rcBuffer & 0x7f800000) >> 23);
            rcBuffer = (rcBuffer & 0x007fffff) << 8;
            rcRange <<= 8;
        }
    }

    function _addWithCarry(value) {
        rcBuffer += value;
        if (rcBuffer > 0x7fffffff) {
            rcBuffer &= 0x7fffffff;
            carry = true;
        }
    }

    function _encodeFlag(probability, match) {
        _normalize();
        var rcHelper = (rcRange >> 15) * probability;
        if (match) {
            rcRange = rcHelper;
        } else {
            _addWithCarry(rcHelper);
            rcRange -= rcHelper;
        }
    }

    function _encodeSkewed(flag) {
        _normalize();
        if (flag) {
            rcRange--;
        } else {
            _addWithCarry(rcRange - 1);
            rcRange = 1;
        }
    }

    function _encodeSingleOnlyLowLzp(nextSymbol) {
        self.computeHashesOnlyLowLzp();
        var lzpStateLow = self.getLzpStateLow();
        var predictedSymbolLow = self.getLzpPredictedSymbolLow();
        var modelLowFrequency = self.getSeeLow(lzpStateLow);
        var matchLow = nextSymbol == predictedSymbolLow;
        _encodeFlag(modelLowFrequency, matchLow);
        self.updateSeeLow(lzpStateLow, matchLow);
        self.updateLzpStateLow(lzpStateLow, nextSymbol, matchLow);
        if (!matchLow) {
            _encodeSymbol(nextSymbol, predictedSymbolLow);
        }
        self.updateContext(nextSymbol);
    }

    function _encodeSingle(nextSymbol) {
        self.computeHashes();
        var lzpStateLow = self.getLzpStateLow();
        var predictedSymbolLow = self.getLzpPredictedSymbolLow();
        var modelLowFrequency = self.getSeeLow(lzpStateLow);
        var lzpStateHigh = self.getLzpStateHigh();
        var predictedSymbolHigh = self.getLzpPredictedSymbolHigh();
        var modelHighFrequency = self.getSeeHigh(lzpStateHigh);
        var matchLow, matchHigh;
        if (modelLowFrequency >= modelHighFrequency) {
            matchHigh = nextSymbol == predictedSymbolHigh;
            self.updateSeeHistoryHigh(matchHigh);
            self.updateLzpStateHigh(lzpStateHigh, nextSymbol, matchHigh);
            matchLow = nextSymbol == predictedSymbolLow;
            _encodeFlag(modelLowFrequency, matchLow);
            self.updateSeeLow(lzpStateLow, matchLow);
            self.updateLzpStateLow(lzpStateLow, nextSymbol, matchLow);
            if (!matchLow) {
                _encodeSymbol(nextSymbol, predictedSymbolLow);
            }
        } else {
            matchLow = nextSymbol == predictedSymbolLow;
            self.updateSeeHistoryLow(matchLow);
            self.updateLzpStateLow(lzpStateLow, nextSymbol, matchLow);
            matchHigh = nextSymbol == predictedSymbolHigh;
            _encodeFlag(modelHighFrequency, matchHigh);
            self.updateSeeHigh(lzpStateHigh, matchHigh);
            self.updateLzpStateHigh(lzpStateHigh, nextSymbol, matchHigh);
            if (!matchHigh) {
                _encodeSymbol(nextSymbol, predictedSymbolHigh);
            }
        }
        self.updateContext(nextSymbol);
    }

    function _encodeSymbol(nextSymbol, mispredictedSymbol) {
        _normalize();
        self.computePpmContext();
        var index = (self.getLastPpmContext() << 8) + nextSymbol;
        var cumulativeExclusiveFrequency = 0;
        var symbolGroup = index >> 4;
        var indexPartial;
        for (indexPartial = self.getLastPpmContext() << 4;
             indexPartial < symbolGroup; indexPartial++) {
            cumulativeExclusiveFrequency +=
                self.getRangesGrouped()[indexPartial];
        }
        for (indexPartial = symbolGroup << 4; indexPartial < index;
             indexPartial++) {
            cumulativeExclusiveFrequency +=
                self.getRangesSingle()[indexPartial];
        }
        var mispredictedSymbolFrequency = self.getRangesSingle()[
            (self.getLastPpmContext() << 8) + mispredictedSymbol];
        if (nextSymbol > mispredictedSymbol) {
            cumulativeExclusiveFrequency -= mispredictedSymbolFrequency;
        }
        var rcHelper = rcRange / (self.getRangesTotal()[
            self.getLastPpmContext()] - mispredictedSymbolFrequency) >> 0;
        _addWithCarry(rcHelper * cumulativeExclusiveFrequency);
        rcRange = rcHelper * self.getRangesSingle()[index];
        self.updatePpm(index);
    }

    self.flush = function() {
        _encodeSkewed(false);
        var i;
        for (i = 0; i < 5; i++) {
            _outputByte((rcBuffer >> 23) & 0xff);
            rcBuffer = (rcBuffer & 0x007fffff) << 8;
        }
    };

    self.encode = function(limit) {
        var endReached = false;
        var i;
        for (i = 0; i < limit; i++) {
            var symbol = inputStream.read();
            if (symbol == -1) {
                endReached = true;
                break;
            }
            _encodeSkewed(true);
            if (self.isOnlyLowLzp()) {
                _encodeSingleOnlyLowLzp(symbol);
            } else {
                _encodeSingle(symbol);
            }
        }
        return endReached;
    };

    return self;
}

File: scalajs/TarsaLZP/app/src/main/scala/pl/tarsa/tarsalzp/compression/engine/Encoder.scala
/*
 * Copyright (C) 2016 Piotr Tarsa ( http://github.com/tarsa )
 *
 *  This software is provided 'as-is', without any express or implied
 *  warranty.  In no event will the author be held liable for any damages
 *  arising from the use of this software.
 *
 *  Permission is granted to anyone to use this software for any purpose,
 *  including commercial applications, and to alter it and redistribute it
 *  freely, subject to the following restrictions:
 *
 *  1. The origin of this software must not be misrepresented; you must not
 *     claim that you wrote the original software. If you use this software
 *     in a product, an acknowledgment in the product documentation would be
 *     appreciated but is not required.
 *  2. Altered source versions must be plainly marked as such, and must not be
 *     misrepresented as being the original software.
 *  3. This notice may not be removed or altered from any source distribution.
 *
 */
package pl.tarsa.tarsalzp.compression.engine

import java.io.{InputStream, OutputStream}

import pl.tarsa.tarsalzp.compression.options.Options


class Encoder(
  inputStream: InputStream,
  outputStream: OutputStream,
  options: Options
) extends Common(options) {

  private var rcBuffer = 0
  private var rcRange = 0x7fffffff
  private var xFFRunLength = 0
  private var lastOutputByte = 0
  private var delay = false
  private var carry = false

  private def outputByte(octet: Int): Unit = {
    if (octet != 0xff || carry) {
      if (delay) {
        outputStream.write(lastOutputByte + (if (carry) 1 else 0))
      }
      while (xFFRunLength > 0) {
        xFFRunLength -= 1
        outputStream.write(if (carry) 0x00 else 0xff)
      }
      lastOutputByte = octet
      delay = true
      carry = false
    } else {
      xFFRunLength += 1
    }
  }

  private def normalize(): Unit = {
    while (rcRange < 0x00800000) {
      outputByte(rcBuffer >> 23)
      rcBuffer = (rcBuffer << 8) & 0x7fffffff
      rcRange <<= 8
    }
  }

  private def addWithCarry(cumulativeExclusiveFraction: Int): Unit = {
    rcBuffer += cumulativeExclusiveFraction
    if (rcBuffer < 0) {
      carry = true
      rcBuffer &= 0x7fffffff
    }
  }

  private def encodeFlag(probability: Int, matched: Boolean): Unit = {
    normalize()
    val rcHelper = (rcRange >> 15) * probability
    if (matched) {
      rcRange = rcHelper
    } else {
      addWithCarry(rcHelper)
      rcRange -= rcHelper
    }
  }

  private def encodeSkewed(flag: Boolean): Unit = {
    normalize()
    if (flag) {
      rcRange -= 1
    } else {
      addWithCarry(rcRange - 1)
      rcRange = 1
    }
  }

  private def encodeSingleOnlyLowLzp(nextSymbol: Int): Unit = {
    computeHashesOnlyLowLzp()
    val lzpStateLow = getLzpStateLow
    val predictedSymbolLow = getLzpPredictedSymbolLow
    val modelLowFrequency = getApmLow(lzpStateLow)
    val matchLow = nextSymbol == predictedSymbolLow
    encodeFlag(modelLowFrequency, matchLow)
    updateApmLow(lzpStateLow, matchLow)
    updateLzpStateLow(lzpStateLow, nextSymbol, matchLow)
    if (!matchLow) {
      encodeSymbol(nextSymbol, predictedSymbolLow)
    }
    updateContext(nextSymbol)
  }

  private def encodeSingle(nextSymbol: Int): Unit = {
    computeHashes()
    val lzpStateLow = getLzpStateLow
    val predictedSymbolLow = getLzpPredictedSymbolLow
    val modelLowFrequency = getApmLow(lzpStateLow)
    val lzpStateHigh = getLzpStateHigh
    val predictedSymbolHigh = getLzpPredictedSymbolHigh
    val modelHighFrequency = getApmHigh(lzpStateHigh)
    if (modelLowFrequency >= modelHighFrequency) {
      val matchHigh = nextSymbol == predictedSymbolHigh
      updateApmHistoryHigh(matchHigh)
      updateLzpStateHigh(lzpStateHigh, nextSymbol, matchHigh)
      val matchLow = nextSymbol == predictedSymbolLow
      encodeFlag(modelLowFrequency, matchLow)
      updateApmLow(lzpStateLow, matchLow)
      updateLzpStateLow(lzpStateLow, nextSymbol, matchLow)
      if (!matchLow) {
        encodeSymbol(nextSymbol, predictedSymbolLow)
      }
    } else {
      val matchLow = nextSymbol == predictedSymbolLow
      updateApmHistoryLow(matchLow)
      updateLzpStateLow(lzpStateLow, nextSymbol, matchLow)
      val matchHigh = nextSymbol == predictedSymbolHigh
      encodeFlag(modelHighFrequency, matchHigh)
      updateApmHigh(lzpStateHigh, matchHigh)
      updateLzpStateHigh(lzpStateHigh, nextSymbol, matchHigh)
      if (!matchHigh) {
        encodeSymbol(nextSymbol, predictedSymbolHigh)
      }
    }
    updateContext(nextSymbol)
  }

  private def encodeSymbol(nextSymbol: Int, mispredictedSymbol: Int): Unit = {
    normalize()
    computeLiteralCoderContext()
    val index = (lastLiteralCoderContext << 8) + nextSymbol
    if (!shouldUseFixedProbabilities) {
      var cumulativeExclusiveFrequency = 0
      val symbolGroup = index >> 4
      for (indexPartial <- (lastLiteralCoderContext << 4) until symbolGroup) {
        cumulativeExclusiveFrequency += rangesGrouped(indexPartial)
      }
      for (indexPartial <- (symbolGroup << 4) until index) {
        cumulativeExclusiveFrequency += rangesSingle(indexPartial)
      }
      val mispredictedSymbolFrequency =
        rangesSingle((lastLiteralCoderContext << 8) + mispredictedSymbol)
      if (nextSymbol > mispredictedSymbol) {
        cumulativeExclusiveFrequency -= mispredictedSymbolFrequency
      }
      val rcHelper = rcRange /
        (rangesTotal(lastLiteralCoderContext) - mispredictedSymbolFrequency)
      addWithCarry(rcHelper * cumulativeExclusiveFrequency)
      rcRange = rcHelper * rangesSingle(index)
    } else {
      rcRange /= 255
      addWithCarry(rcRange *
        (nextSymbol - (if (nextSymbol > mispredictedSymbol) 1 else 0)))
    }
    updateRecentCost(rangesSingle(index), rangesTotal(lastLiteralCoderContext))
    updateLiteralCoder(index)
  }

  def flush(): Unit = {
    for (i <- 0 until 5) {
      outputByte((rcBuffer >> 23) & 0xff)
      rcBuffer <<= 8
    }
  }

  def encode(limit: Int): Int = {
    var endReached = false
    var result = limit
    for (i <- 0 until limit if !endReached) {
      val symbol = inputStream.read()
      endReached = symbol == -1
      encodeSkewed(!endReached)
      if (endReached) {
        result = i
      } else {
        if (onlyLowLzp) {
          encodeSingleOnlyLowLzp(symbol)
        } else {
          encodeSingle(symbol)
        }
      }
    }
    result
  }
}
